clang/test/SemaCXX/cheri/cheri-int-to-cap-cast.cpp

```diff
@@ -0,0 +1,129 @@
+// purecap-note@-4 {{in instantiation of member function 'TestBasic<long, void *>::test_reinterpret_cast' requested here}}
+// purecap-note@-5 {{in instantiation of member function 'TestBasic<long, void *>::test_dynamic_cast' requested here}}
+
+template <typename S, typename T>
+class TestQualified {
+  T __capability test_static_cast(S s) {
+    // static_cast isn't valid for integer to pointer casts.
+    return static_cast<T __capability>(s);
+    // hybrid-error@-1 {{cannot cast from type 'int' to pointer type 'void * __capability'}}
+    // purecap-error@-2 {{cannot cast from type 'int' to pointer type 'void *'}}
+    // hybrid-error@-3 {{cannot cast from type 'long' to pointer type 'void * __capability'}}
+    // purecap-error@-4 {{cannot cast from type 'long' to pointer type 'void *'}}
+  }
+
+  T __capability test_reinterpret_cast(S s) {
+    // These are all currently fine for hybrid.
+    return reinterpret_cast<T __capability>(s);
+    // purecap-warning@-1 2 {{cast from provenance-free integer type to pointer type will give pointer that can not be dereferenced}}
+  }
+
+  T __capability test_dynamic_cast(S s) {
+    // dynamic_cast isn't valid for integer to pointer casts.
+    return dynamic_cast<T __capability>(s);
+    // hybrid-error@-1 {{cannot use dynamic_cast to convert from 'int' to 'void * __capability'}}
+    // purecap-error@-2 {{cannot use dynamic_cast to convert from 'int' to 'void *'}}
+    // hybrid-error@-3 {{cannot use dynamic_cast to convert from 'long' to 'void * __capability'}}
+    // purecap-error@-4 {{cannot use dynamic_cast to convert from 'long' to 'void *'}}
+  }
+};
+
+template class TestQualified<int, void *>;
+// hybrid-note@-1 {{in instantiation of member function 'TestQualified<int, void *>::test_static_cast' requested here}}
+// hybrid-note@-2 {{in instantiation of member function 'TestQualified<int, void *>::test_dynamic_cast' requested here}}
+// purecap-note@-3 {{in instantiation of member function 'TestQualified<int, void *>::test_static_cast' requested here}}
+// purecap-note@-4 {{in instantiation of member function 'TestQualified<int, void *>::test_reinterpret_cast' requested here}}
+// purecap-note@-5 {{in instantiation of member function 'TestQualified<int, void *>::test_dynamic_cast' requested here}}
+template class TestQualified<long, void *>;
+// hybrid-note@-1 {{in instantiation of member function 'TestQualified<long, void *>::test_static_cast' requested here}}
+// hybrid-note@-2 {{in instantiation of member function 'TestQualified<long, void *>::test_dynamic_cast' requested here}}
+// purecap-note@-3 {{in instantiation of member function 'TestQualified<long, void *>::test_static_cast' requested here}}
+// purecap-note@-4 {{in instantiation of member function 'TestQualified<long, void *>::test_reinterpret_cast' requested here}}
+// purecap-note@-5 {{in instantiation of member function 'TestQualified<long, void *>::test_dynamic_cast' requested here}}
```

Comment:
```suggestion
// purecap-note@-5 {{in instantiation of member function 'TestQualified<long, void *>::test_dynamic_cast' requested here}}
template class TestQualified<void *, void *>;
```

Shouldn't we also check `void *` -> `void *__capability` casts?

Reply:
Those are done in cheri-cap-ptr-casts.cpp as they're not language-level integers, as well as some of the tests I added recently with all the `__capability` qualifier cleanups and fixes.